group-generators/helpers/data-providers/rep3/index.ts

```diff
@@ -0,0 +1,97 @@
+import { gql } from "graphql-request";
+import { QueryMembersOutput, QueryMembersInput, IRep3Provider } from "./types";
+
+import { GraphQLProvider } from "@group-generators/helpers/data-providers/graphql";
+import { FetchedData } from "topics/group";
+
+export class Rep3Provider extends GraphQLProvider implements IRep3Provider {
+  constructor() {
+    super({
+      url: "https://api.thegraph.com/subgraphs/name/eth-jashan/rep3-matic",
+    });
+  }
+
+  public async getMembershipHolders({
+    contract,
+    level,
+  }: QueryMembersInput): Promise<FetchedData> {
+    const holders: FetchedData = {};
+
+    if (!this.isValidLevel(level)) {
+      throw new Error(
+        "Invalid level, has to be `all` or a number between 1 and 256"
+      );
+    }
+    const levelSetting =
+      level === "all" ? `level_gte: "1"` : `level: "${level}"`;
+
+    try {
+      const pageSize = 1000;
+      let skip = 0;
+      let continuePaging = true;
+
+      while (continuePaging) {
+        const response: QueryMembersOutput = await this.fetchPage(
+          contract,
+          levelSetting,
+          pageSize,
+          skip
+        );
+
+        if (response.membershipNFTs.length > 0) {
+          response.membershipNFTs.forEach((holder) => {
+            holders[holder.claimer] = 1;
```

Comment:
It could be nice to put the level we chose as the value of the address => quick win (not optimized) method

But it could be even better to put the real level of the address in value. (e.g. if we request all levels, for each address, we should have is associated level in value) => best method

Reply:
Thanks @MartinGbz  great suggestion. I updated the PR to include the level